src/core_save.c

```diff
@@ -13,299 +13,147 @@
+    int32_t status = EXIT_FAILURE;
+
+    // Reset the output galaxy count.
+    int32_t OutputGalCount[run_params->MAXSNAPS];
+    for(int32_t snap_idx = 0; snap_idx < run_params->MAXSNAPS; snap_idx++) {
+        OutputGalCount[snap_idx] = 0;
     }
 
-    // track the order in which galaxies are written
-    int *OutputGalOrder = calloc(numgals, sizeof(OutputGalOrder[0]));
+    // Track the order in which galaxies are written.
+    int32_t *OutputGalOrder = calloc(numgals, sizeof(*(OutputGalOrder)));
     if(OutputGalOrder == NULL) {
         fprintf(stderr,"Error: Could not allocate memory for %d int elements in array `OutputGalOrder`\n", numgals);
-        ABORT(MALLOC_FAILURE);
+        return MALLOC_FAILURE;
     }
 
-    for(int i = 0; i < numgals; i++) {
-        OutputGalOrder[i] = -1;
-        haloaux[i].output_snap_n = -1;
+    for(int32_t gal_idx = 0; gal_idx < numgals; gal_idx++) {
+        OutputGalOrder[gal_idx] = -1;
+        haloaux[gal_idx].output_snap_n = -1;
     }
-  
-    // first update mergeIntoID to point to the correct galaxy in the output
-    for(int n = 0; n < run_params->NOUT; n++) {
-        for(int i = 0; i < numgals; i++) {
-            if(halogal[i].SnapNum == run_params->ListOutputSnaps[n]) {
-                OutputGalOrder[i] = OutputGalCount[n];
-                OutputGalCount[n]++;
-                haloaux[i].output_snap_n = n;
+
+    // First update mergeIntoID to point to the correct galaxy in the output.
+    for(int32_t snap_idx = 0; snap_idx < run_params->NOUT; snap_idx++) {
+        for(int32_t gal_idx  = 0; gal_idx < numgals; gal_idx++) {
+            if(halogal[gal_idx].SnapNum == run_params->ListOutputSnaps[snap_idx]) {
+                OutputGalOrder[gal_idx] = OutputGalCount[snap_idx];
+                OutputGalCount[snap_idx]++;
+                haloaux[gal_idx].output_snap_n = snap_idx;
             }
         }
     }
 
-    int num_output_gals = 0;
-    int num_gals_processed[run_params->MAXSNAPS];
-    memset(num_gals_processed, 0, sizeof(num_gals_processed[0])*run_params->MAXSNAPS);
-    for(int n = 0; n < run_params->NOUT; n++) {
-        cumul_output_ngal[n] = num_output_gals;
-        num_output_gals += OutputGalCount[n];
-    }
-    
-    for(int i = 0; i < numgals; i++) {
-        if(halogal[i].mergeIntoID > -1) {
-            halogal[i].mergeIntoID = OutputGalOrder[halogal[i].mergeIntoID];
+    for(int32_t gal_idx = 0; gal_idx < numgals; gal_idx++) {
+        if(halogal[gal_idx].mergeIntoID > -1) {
+            halogal[gal_idx].mergeIntoID = OutputGalOrder[halogal[gal_idx].mergeIntoID];
         }
     }
 
-    struct GALAXY_OUTPUT *all_outputgals  = calloc(num_output_gals, sizeof(struct GALAXY_OUTPUT));
-    if(all_outputgals == NULL) {
-        fprintf(stderr,"Error: Could not allocate enough memory to hold all %d output galaxies\n",num_output_gals);
-        ABORT(MALLOC_FAILURE);
-    }
+    // All of the tracking arrays set up, time to perform the actual writing.
```

Comment:
I've added the following code snippet:

```
for(int32_t gal_idx = 0; gal_idx < numgals; gal_idx++) {
  generate_galaxy_index(<All the necessary parameters to pass>, &halogal[gal_idx].GalaxyIndex);
}
```

And I'm updating the `halogal[gal_idx].GalaxyIndex` as a pointer inside the `generate_galaxy_index` function.

We also need to add a `CentralGalaxyIndex` value; which is simply the `GalaxyIndex` value for the Central Galaxy for `halogal[gal_idx]`'s FoF group.

My question, is should we handle `CentralGalaxyIndex` inside `generate_galaxy_index`?  That is, should I also pass `&halogal[gal_idx].CentralGalaxyIndex` and then calculate it using something like `halogal[haloaux[halos[g->HaloNr].FirstHaloInFOFgroup].FirstGalaxy].GalaxyNr + TREE_MUL_FAC * treenr + (THISTASK_MUL_FAC/10) * filenr`?  

Or should I instead do a secondary loop outside where I access the value directly:

```
// Continuing on from the above snippet...
for(int32_t gal_idx = 0; gal_idx < numgals; gal_idx++) {
  int64_t CentralGalaxyIndex = halogal[haloaux[halos[g->HaloNr].FirstHaloInFOFgroup].FirstGalaxy].GalaxyIndex;
  halogal[gal_idx].CentralGalaxyIndex = CentralGalaxyIndex;
}
```

Reply:
The loop should be inside the function. Otherwise, change the function to be a macro.

Just handle both of them inside the function; however, remember that there is no guarantee that the `CentralGalaxyIndex` will already have been assigned (as in, you shouldn't assume that `centralgal` occurs before the `subhalo`). Another option is that **since** you already know `centralgal`, you can similarly compute the `CentralGalaxyIndex` while assigning `GalaxyIndex`. This is wasteful in terms of compute but certainly better than looping once again through all the galaxies.